File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        if (x >= y) {
            return removeSubstrings(s, 'a', 'b', x) + removeSubstrings(s, 'b', 'a', y);
        } else {
            return removeSubstrings(s, 'b', 'a', y) + removeSubstrings(s, 'a', 'b', x);
        }
    }

private:
    int removeSubstrings(string& s, char first, char second, int points) {
        string result;
        int score = 0;

        for (char c : s) {
            if (!result.empty() && result.back() == first && c == second) {
                result.pop_back(); // Remove the matching pair
                score += points;   // Gain points
            } else {
                result.push_back(c); // Push current character onto result
            }
        }

        s = result; // Update s to the remaining characters after removals
        return score;
    }
};

File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        if x >= y:
            s, score1 = self.removeSubstrings(s, 'a', 'b', x)
            s, score2 = self.removeSubstrings(s, 'b', 'a', y)
        else:
            s, score1 = self.removeSubstrings(s, 'b', 'a', y)
            s, score2 = self.removeSubstrings(s, 'a', 'b', x)
        return score1 + score2

    def removeSubstrings(self, s: str, first: str, second: str, points: int) -> (str, int):
        stack = []
        score = 0

        for char in s:
            if stack and stack[-1] == first and char == second:
                stack.pop()  # Remove the matching pair
                score += points  # Gain points
            else:
                stack.append(char)  # Push current character onto stack

        # Convert the stack back to string
        return ''.join(stack), score